File: 1453.maximum-number-of-darts-inside-of-a-circular-dartboard.py
#
# @lc app=leetcode id=1453 lang=python3
#
# [1453] Maximum Number of Darts Inside of a Circular Dartboard
#

# @lc code=start
from math import acos, atan2, sqrt


class Solution:
    # From each dart, sweep a circle of radius r, with the dart lying on the circumference
    # Check for the other darts, their distance d: 
    # if d <= 2*r, that means an angular sweep from the first dart would include the second dart at some point.
    # For such second darts, we record their angle of entrance and leave the rotating circle
    # After going all such candidates, we sort the angles of entrance and leave
    # Record the maximum number of darts seen so far 
    def numPoints(self, darts: List[List[int]], r: int) -> int:
        result = 0
        for x, y in darts:
            maxsofar, current = 1, 1
            angles = []
            for x1, y1 in darts:
                if x1 != x or y1 != y:
                    d = sqrt((x1-x)*(x1-x) + (y1-y)*(y1-y))
                    if d <= 2*r:
                        theta = atan2(x1-x,y1-y)
                        delta = acos(d / 2 / r)
                        angles.append([theta - delta, 1])
                        angles.append([theta + delta, -1])
            angles.sort(key = lambda x: [x[0], -x[1]])

            for angle, state in angles:
                current += state
                maxsofar = max(maxsofar, current)
            result = max(result, maxsofar)
        return result
        
# @lc code=end

File: 1453.maximum-number-of-darts-inside-of-a-circular-dartboard.cs
/*
 * @lc app=leetcode id=1453 lang=csharp
 *
 * [1453] Maximum Number of Darts Inside of a Circular Dartboard
 */

// @lc code=start
public class Solution 
{
    public int NumPoints(int[][] darts, int r) 
    {
        int result = 0;
        int n = darts.Length;
        for (int i = 0; i < n; i++)
        {
            int x = darts[i][0];
            int y = darts[i][1];
            int maxsofar = 1;
            int current = 1;
            List<double[]> angles = new List<double[]>();
            for (int j = 0; j < n; j++)
            {
                if (i != j)
                {
                    int x1 = darts[j][0];
                    int y1 = darts[j][1];
                    double d = Math.Sqrt((x1-x)*(x1-x) + (y1-y)*(y1-y));
                    if (d <= (double) 2*r)
                    {
                        double theta = Math.Atan2((double) (x1-x), (double) (y1-y));
                        double delta = Math.Acos(d / 2 / (double) r);
                        angles.Add(new double[2] {theta - delta, 1});
                        angles.Add(new double[2] {theta + delta, -1});
                    }
                }
            }
            angles = angles.OrderBy(t => t[0]).ThenBy( t => - t[1]).ToList();

            foreach (double[] angle in angles)
            {
                current += (int) angle[1];
                maxsofar = Math.Max(maxsofar, current);
            }
            result = Math.Max(result, maxsofar);
        }
        return result;
    }
}
// @lc code=end

